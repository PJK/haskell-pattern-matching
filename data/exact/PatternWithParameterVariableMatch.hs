module PatternWithParameterVariableMatch where

data MyTrinary = A | B | C

extract :: Maybe MyTritary -> Maybe MyTritary -> MyTrinary
extract (Just C) (Just A) = A
extract (Just x)  y       = B -- This should catch Just A, Just B, _
extract  Nothing  x       = B


partialExtract :: Maybe MyTritary -> Maybe MyTritary -> MyTrinary
partialExtract (Just C) (Just A) = A
partialExtract (Just x)  y       = B -- This should catch Just A, Just B, _
partialExtract  Nothing (Just C) = B


