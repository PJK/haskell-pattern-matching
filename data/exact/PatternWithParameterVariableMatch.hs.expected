{
    "tag": "AnalysisSuccess",
    "contents": [
        [
            "partialExtract",
            {
                "tag": "NonExhaustive",
                "contents": [
                    [
                        [
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Nothing",
                                    []
                                ]
                            },
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Nothing",
                                    []
                                ]
                            }
                        ],
                        ""
                    ],
                    [
                        [
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Nothing",
                                    []
                                ]
                            },
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Just",
                                    [
                                        {
                                            "tag": "ConstructorPattern",
                                            "contents": [
                                                "A",
                                                []
                                            ]
                                        }
                                    ]
                                ]
                            }
                        ],
                        ""
                    ],
                    [
                        [
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Nothing",
                                    []
                                ]
                            },
                            {
                                "tag": "ConstructorPattern",
                                "contents": [
                                    "Just",
                                    [
                                        {
                                            "tag": "ConstructorPattern",
                                            "contents": [
                                                "B",
                                                []
                                            ]
                                        }
                                    ]
                                ]
                            }
                        ],
                        ""
                    ]
                ]
            }
        ]
    ]
}
